src/node_v8.cc

```diff
@@ -389,6 +389,72 @@ void GCProfiler::Stop(const FunctionCallbackInfo<v8::Value>& args) {
                                 .ToLocalChecked());
 }
 
+static void CostBeforeGC(Isolate* isolate,
+                         v8::GCType gc_type,
+                         v8::GCCallbackFlags flags,
+                         void* data) {
+  Environment* env = static_cast<Environment*>(data);
+  if (env->current_gc_type() != 0) {
+    return;
+  }
+  env->set_current_gc_type(gc_type);
+  env->set_gc_start_time(uv_hrtime());
+}
+
+static void CostAfterGC(Isolate* isolate,
+                        v8::GCType gc_type,
+                        v8::GCCallbackFlags flags,
+                        void* data) {
+  Environment* env = static_cast<Environment*>(data);
+  if (env->current_gc_type() != gc_type) {
+    return;
+  }
+  env->set_current_gc_type(0);
+  env->set_gc_time(uv_hrtime() - env->gc_start_time());
+  env->set_gc_start_time(0);
+}
+
+void CleanupGCCallback(void* data) {
+  Environment* env = static_cast<Environment*>(data);
+  env->isolate()->RemoveGCPrologueCallback(CostBeforeGC, data);
+  env->isolate()->RemoveGCEpilogueCallback(CostAfterGC, data);
+}
+
+void SubscribeGCCost(const FunctionCallbackInfo<Value>& args) {
+  Environment* env = Environment::GetCurrent(args);
+  env->set_gc_ref(env->gc_ref() + 1);
```

Comment:
It seems weird how this ref is implemented - so if somewhere someone unsubscribes more than once, there is no way to find out if they are unsubscribing the subscribers they added, or if they intend to cancel subscribers added by other code? But if we want to have a proper reference for the subscriber, I think it really would be better to just merge this into the GCProfiler API (e.g. if the user passes an option that specifies only the accumulative time should be tracked, we only track the accumulative time and do not write/buffer the JSON). Unlike CPU usage, GC time tracking has to be explicitly enabled and disabled, and introducing another API that basically is implemented in the same way as GCProfiler seems too complicated.

Reply:
Thanks, Let me think a little bit more.